cpp/src/arrow/util/future_test.cc

```diff
@@ -768,4 +1361,27 @@ TYPED_TEST(FutureIteratorTest, ErrorsAsCompleted) { this->TestErrorsAsCompleted(
 
 TYPED_TEST(FutureIteratorTest, StressAsCompleted) { this->TestStressAsCompleted(); }
 
+namespace internal {
+TEST(FnOnceTest, MoveOnlyDataType) {
+  // ensuring this is valid guarantees we are making no unnecessary copies
```

Comment:
Nice. Do we want to use `FnOnce` in other places at some point? (perhaps open a JIRA?)

Reply:
Yes, I think it should replace `std::function` in `thread_pool.h` at least

Reply:
https://issues.apache.org/jira/browse/ARROW-10966